Handle case of existing package-info.java file correctly and pick up …

tobiasstamann opened the pull request:
…even not annotated TypeElements via Packages.

This still won’t guarantee that all TypeElement will be used. But ensures that only one annotation must be present per package.

tobiasstamann commented:
Hi, if a package-info. java  file is present in package the current implementation wouldn't work since getRootElements would provide the PackageElement instead of the TypeElement.
The fix moves validation to processingDone flagged wave and collects all Package names it can get first to then collect all types.
This change should also enable checks on nested classes.

tobiasstamann commented:
The code won't currently validate nested classes. I am going to add that as well.

odrotbohm commented:
Any reason you tweaked the package name of the example classes? They reside in the `example` package and thus should have this declared as package, too? Is there any chance you avoid the reformatting? If that gets too much of a hassle, never mind. I can easily fix that myself during the merge.

tobiasstamann commented:
Hi, can you please wait with merge until i have added support for nested classes?

I have moved the example classes to another package because some IDEs are compiling even java files in resource folders. 
Cute offers a way to circumvent that. You can name test file *.java.ct to prohibit this, but this has the tradeoff that most autocompletion in IDEs wont work. So best way is to have kind of /testcases/testcaseX/ structure. The compiler doesn't need the folder structure match the package structure.  Just the ClassName must match the file name.

With my implementation the processor would scan for example package names , then it would pick up the compiled classes from example folder.  That's why there could be a hassle.

tobiasstamann commented:
Do you provide have some kind default formatting/style config i can use?

tobiasstamann commented:
Validation of nested classes is now supported as well.

odrotbohm commented:
I've folded your changes into GH-230. I had to move the valid probes for the test cases into a separate packages as otherwise, the test cases checking those valid files now included the invalid ones producing errors and thus broke the tests.
